### PR TITLE
#22813 Create end point to determine if the user gets or not into a E…

### PR DESCRIPTION
**Pull request by freddyucv**
Create a new end point and WebAPI to determine if a user should go into a Experiment

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.dotcms.experiments.business.web;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.io.Serializable;
+
+public  class SelectedVariant implements Serializable {
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java:6]: Missing Javadoc.

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

File: dotCMS/src/curl-test/Experiments Resource.postman_collection.json [hunk 2]
```diff
@@ -512,6 +512,808 @@
 				}
 			]
 		},
+		{
+			"name": "isUserIncluded",
```
Comment (dsilvam) [inline dotCMS/src/curl-test/Experiments Resource.postman_collection.json:516]: Put these tests inside a new folder Called "isUserIncluded"

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPI.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+     * - First it checks it the {@link Experiment#targetingConditions()} is valid for the user or current
+     * {@link HttpServletRequest}.
+     * - Then it use the {@link Experiment#trafficAllocation()} to know if the user should go into the
+     * {@link Experiment}.
+     * - Finally it assing a {@link com.dotcms.experiments.model.ExperimentVariant} according to
+     * {@link com.dotcms.experiments.model.ExperimentVariant#weight()}
+     *
+     * If exists more that one {@link Experiment} RUNNING it try to get the user into any of them
+     * one by one if finally the user is not going into any experiment then it return a
+     * {@link com.dotcms.experiments.business.web.ExperimentWebAPI#NONE_EXPERIMENT}
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPI.java:29]: I think we need to support the same user in multiple experiments... that is, experiments for different pages

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import net.bytebuddy.utility.RandomString;
+import org.jetbrains.annotations.NotNull;
+
+public class ExperimentWebAPIImpl implements ExperimentWebAPI {
```
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java:31]: It's at API level @jcastro-dotcms

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/experiments/ExperimentsResource.java [hunk 5]
```diff
@@ -359,6 +360,35 @@ public ResponseEntitySingleExperimentView deleteTargetingCondition(@Context fina
...
+     * - First it checks it the {@link Experiment#targetingConditions()} is valid for the user or current
+     * {@link HttpServletRequest}.
+     * - Then it use the {@link Experiment#trafficAllocation()} to know if the user should go into the
+     * {@link Experiment}.
+     * - Finally it assing a {@link com.dotcms.experiments.model.ExperimentVariant} according to
+     * {@link com.dotcms.experiments.model.ExperimentVariant#weight()}
+     *
+     * If exists more that one {@link Experiment} RUNNING it try to get the user into any of them
+     * one by one if finally the user is not going into any experiment then it return a
+     * {@link com.dotcms.experiments.business.web.ExperimentWebAPI#NONE_EXPERIMENT}
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/experiments/ExperimentsResource.java:375]: same discussion

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java [hunk 6]
```diff
@@ -114,4 +116,21 @@ Experiment deleteVariant(String experimentId, String variantName, User user)
 
     Experiment deleteTargetingCondition(String experimentId, String conditionId, User user)
             throws DotDataException, DotSecurityException;
+
+    /**
+     * Return a list of the current RUNNING Experiments.
+     *
+     * @return
+     */
+    List<Experiment> getExperimentRunning() throws DotDataException;
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java:125]: rename to "getRunningExperiments"

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/experiments/ResponseEntityExperimentSelectedView.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package com.dotcms.rest.api.v1.experiments;
+
+import com.dotcms.experiments.business.web.SelectedExperiment;
+import com.dotcms.rest.ResponseEntityView;
+
+public class ResponseEntityExperimentSelectedView extends
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/experiments/ResponseEntityExperimentSelectedView.java:6]: javadoc

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
+package com.dotcms.experiments.business.web;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.io.Serializable;
+
+public class SelectedExperiment implements Serializable {
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java:15]: https://github.com/dotCMS/core/pull/23334/commits/66edbb70b1ba0fc540f11e43f0ba163c6cacf4e1#diff-282d55898ddcc13745c364b2a7277eea0babbd5ab656e4100748e1a886ece800R8

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java:37]: done https://github.com/dotCMS/core/pull/23334/commits/66edbb70b1ba0fc540f11e43f0ba163c6cacf4e1#diff-df4c3ee11b473e0f4e91d5be7a6107bf6cdb4ab40bf0661a182dd298e00e5602R31

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+        for (ExperimentVariant variant : trafficProportion.variants()) {
+            totalWeight += variant.weight();
+
+            if (randomValue < totalWeight) {
+                return new SelectedVariant(variant.id(), "");
+            }
+        }
+
+        throw new IllegalStateException("Should return one variant");
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java]: done https://github.com/dotCMS/core/pull/23334/commits/66edbb70b1ba0fc540f11e43f0ba163c6cacf4e1#diff-df4c3ee11b473e0f4e91d5be7a6107bf6cdb4ab40bf0661a182dd298e00e5602R106

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java:15]: done https://github.com/dotCMS/core/pull/23334/commits/66edbb70b1ba0fc540f11e43f0ba163c6cacf4e1#diff-f14f138414a0e78f2a8cdaeaa998426d2b3c35639d67b8dcd420baa5f2714d5aR8

File: dotCMS/src/curl-test/Experiments Resource.postman_collection.json (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/curl-test/Experiments Resource.postman_collection.json]: done https://github.com/dotCMS/core/pull/23334/commits/df9d28d9d0691423d48c1e0ab8308ef130e4e993

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPI.java (hunk 3 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPI.java:30]: done https://github.com/dotCMS/core/pull/23334/commits/8b54e850f9195005465ca1899d4d62a5cd72d7e8

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java (hunk 6 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPI.java]: done https://github.com/dotCMS/core/pull/23334/commits/5c94901d63ceb1249a7832d916be5be6f3b380dd

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/experiments/ResponseEntityExperimentSelectedView.java (hunk 7 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/experiments/ResponseEntityExperimentSelectedView.java:11]: done https://github.com/dotCMS/core/pull/23334/commits/516f20f9655686e60ea7e01cfc5490a142aa539f#diff-69f2d16140d00ac7ea312aa4c96f2fd9513e1b02165a88606921720b7ef9d4ebR8

**Review by jcastro-dotcms — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 10]
```diff
@@ -488,6 +489,22 @@ public Experiment deleteTargetingCondition(String experimentId, String condition
 
     }
 
+    @Override
+    public List<Experiment> getRunningExperiment() throws DotDataException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:493]: Close Db annotation?

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                        .map(experiment -> createSelectedExperimentAndCreateCookie(response,
+                                experiment))
+                        .collect(Collectors.toList());
+        } else {
+            setCookie(NONE_EXPERIMENT, response, getDefaultExpireDate());
+            return list(NONE_EXPERIMENT);
+        }
+    }
+
+    @NotNull
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java:58]: remove this

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        runningExperimentCookie.setMaxAge((int) res.getSeconds());
+
+        response.addCookie(runningExperimentCookie);
+    }
+
+    private String getCookieName(final SelectedExperiment experimentSelected) {
+        return "runningExperiment_" + experimentSelected.id();
+    }
+
+    @NotNull
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java:93]: remove

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    @JsonProperty("pageUrl")
+    private String pageUrl;
+
+    /**
+     * Selected Variant for the User
+     */
+    @JsonProperty("variant")
+    private SelectedVariant variant;
+
+    public SelectedExperiment(final String id, final String name, final String pageUrl,
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java:41]: does not need the JsonCreator annotation?

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+     * Experiment's ID
+     */
+    @JsonProperty("id")
+    private String id;
+
+    /**
+     * Experiment;s page URL
+     */
+    @JsonProperty("pageUrl")
+    private String pageUrl;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java:33]: set all of them to final

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    @JsonProperty("name")
+    private String name;
+
+    /**
+     * URL to render the {@link com.dotcms.experiments.model.Experiment}'s page in this {@link com.dotcms.variant.model.Variant}
+     */
+    @JsonProperty("url")
+    private String url;
+
+    public SelectedVariant(String name, String url) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java:29]: same set to final and JsonCreator

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java (hunk 10 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:495]: done https://github.com/dotCMS/core/pull/23334

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java (hunk 11 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java]: done https://github.com/dotCMS/core/pull/23334

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java (hunk 12 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/ExperimentWebAPIImpl.java]: done https://github.com/dotCMS/core/pull/23334/commits/088b10e51e9d5864fb510bd1a7cc17241899e47e#diff-df4c3ee11b473e0f4e91d5be7a6107bf6cdb4ab40bf0661a182dd298e00e5602L93

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java (hunk 13 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java:41]: No, it is working without it

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java (hunk 14 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedExperiment.java]: done https://github.com/dotCMS/core/pull/23334/commits/088b10e51e9d5864fb510bd1a7cc17241899e47e#diff-282d55898ddcc13745c364b2a7277eea0babbd5ab656e4100748e1a886ece800R21

File: dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java (hunk 15 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/business/web/SelectedVariant.java:29]: done https://github.com/dotCMS/core/pull/23334/commits/088b10e51e9d5864fb510bd1a7cc17241899e47e#diff-f14f138414a0e78f2a8cdaeaa998426d2b3c35639d67b8dcd420baa5f2714d5aR21

**Review by jdotcms — APPROVED** [review]

**Pull request merged**